jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/ProjectCommand.java

```diff
@@ -79,4 +96,79 @@ private static String getWorkspaceInvisibleProjectName(IPath workspacePath) {
         String fileName = workspacePath.toFile().getName();
         return fileName + "_" + Integer.toHexString(workspacePath.toPortableString().hashCode());
     }
+
+    public static List<PackageNode> getMainMethod(IProgressMonitor monitor) {
+        final List<PackageNode> res = new ArrayList<>();
+        IJavaSearchScope scope = SearchEngine.createWorkspaceScope();
+        SearchPattern pattern = SearchPattern.createPattern("main(String[]) void", IJavaSearchConstants.METHOD,
+                IJavaSearchConstants.DECLARATIONS, SearchPattern.R_EXACT_MATCH| SearchPattern.R_CASE_SENSITIVE);
+        SearchRequestor requestor = new SearchRequestor() {
+            @Override
+            public void acceptSearchMatch(SearchMatch match) {
+                Object element = match.getElement();
+                if (element instanceof IMethod) {
+                    IMethod method = (IMethod) element;
+                    try {
+                        if (method.isMainMethod()) {
+                            IResource resource = method.getResource();
+                            if (resource != null) {
+                                IProject project = resource.getProject();
+                                if (project != null) {
+                                    String mainClass = method.getDeclaringType().getFullyQualifiedName();
+                                    IJavaProject javaProject = getJavaProject(project);
```

Comment:
Why we do not directly call `method.getJavaProject()` here?

Reply:
> Why we do not directly call `method.getJavaProject()` here?

It's better to use `method.getJavaProject()`.